saleor/discount/utils/voucher.py

```diff
@@ -304,22 +325,45 @@ def get_products_voucher_discount(
     return total_amount
 
 
+def create_or_update_voucher_discount_objects_for_order(
+    order: "Order", denormalized=False
+):
+    from ...order.fetch import fetch_draft_order_lines_info
+    from ...order.models import OrderLine
+
+    create_or_update_discount_object_from_order_level_voucher(order)
+    lines_info = fetch_draft_order_lines_info(order)
+    create_or_update_line_discount_objects_from_voucher(lines_info, denormalized)
+    lines = [line_info.line for line_info in lines_info]
+    OrderLine.objects.bulk_update(lines, ["base_unit_price_amount"])
+
+
 def create_or_update_discount_object_from_order_level_voucher(
-    order, database_connection_name
+    order,
+    database_connection_name: str = settings.DATABASE_CONNECTION_DEFAULT_NAME,
 ):
     """Create or update discount object for ENTIRE_ORDER and SHIPPING voucher."""
     voucher = order.voucher
-    if not order.voucher_id or (
-        is_order_level_voucher(voucher)
-        and order.discounts.filter(type=DiscountType.MANUAL)
-    ):
+
+    # The order-level voucher discount should be deleted when:
+    # - order.voucher_id is None
+    # - manual order-level discount exists together with order-level voucher
+    # - the order has line-level voucher attached
+    should_delete_order_level_voucher_discount = (
+        not order.voucher_id
+        or (
+            is_order_level_voucher(voucher)
+            and order.discounts.filter(type=DiscountType.MANUAL)
+        )
+        or (not is_order_level_voucher(voucher) and not is_shipping_voucher(voucher))
+    )
```

Comment:
Let's improve readability
```suggestion
    is_manual_discount = order.discounts.filter(type=DiscountType.MANUAL).exists()
    is_order_voucher = is_order_level_voucher(voucher)
    is_line_level_voucher = (not is_order_voucher and not  is_shipping_voucher(voucher))
    
    should_delete_order_level_voucher_discount = (
        not order.voucher_id
        or (is_order_voucher and is_manual_discount)
        or is_line_level_voucher
    )
    
```

Reply:
done as requested